pkg/blobcache/blobcache.go

```diff
@@ -0,0 +1,571 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+)
+
+var (
+	_ types.ImageReference   = &blobCacheReference{}
+	_ types.ImageSource      = &blobCacheSource{}
+	_ types.ImageDestination = &blobCacheDestination{}
+)
+
+const (
+	compressedNote   = ".compressed"
+	decompressedNote = ".decompressed"
+)
+
+// BlobCache is an object which saves copies of blobs that are written to it while passing them
+// through to some real destination, and which can be queried directly in order to read them
+// back.
+type BlobCache interface {
+	types.ImageReference
+	// HasBlob checks if a blob that matches the passed-in digest (and
+	// size, if not -1), is present in the cache.
+	HasBlob(types.BlobInfo) (bool, int64, error)
+	// Directories returns the list of cache directories.
+	Directory() string
+	// ClearCache() clears the contents of the cache directories.  Note
+	// that this also clears content which was not placed there by this
+	// cache implementation.
+	ClearCache() error
+}
+
+type blobCacheReference struct {
+	reference types.ImageReference
+	// WARNING: The contents of this directory may be accessed concurrently,
+	// both within this process and by multiple different processes
+	directory string
+	compress  types.LayerCompression
+}
+
+type blobCacheSource struct {
+	reference *blobCacheReference
+	source    types.ImageSource
+	sys       types.SystemContext
+	// this mutex synchronizes the counters below
+	mu          sync.Mutex
+	cacheHits   int64
+	cacheMisses int64
+	cacheErrors int64
+}
+
+type blobCacheDestination struct {
+	reference   *blobCacheReference
+	destination types.ImageDestination
+}
+
+func makeFilename(blobSum digest.Digest, isConfig bool) string {
+	if isConfig {
+		return blobSum.String() + ".config"
+	}
+	return blobSum.String()
+}
+
+// CacheLookupReferenceFunc wraps a BlobCache into a
+// types.LookupReferenceFunc to allow for using a BlobCache during
+// image-copy operations.
+func CacheLookupReferenceFunc(directory string, compress types.LayerCompression) types.LookupReferenceFunc {
+	// NOTE: this prevents us from moving BlobCache around and generalizes
+	// the libimage API.
+	return func(ref types.ImageReference) (types.ImageReference, error) {
+		ref, err := NewBlobCache(ref, directory, compress)
+		if err != nil {
+			return nil, errors.Wrapf(err, "error using blobcache %q", directory)
+		}
+		return ref, nil
+	}
+}
+
+// NewBlobCache creates a new blob cache that wraps an image reference.  Any blobs which are
+// written to the destination image created from the resulting reference will also be stored
+// as-is to the specified directory or a temporary directory.  The cache directory's contents
+// can be cleared by calling the returned BlobCache()'s ClearCache() method.
+// The compress argument controls whether or not the cache will try to substitute a compressed
+// or different version of a blob when preparing the list of layers when reading an image.
```

Comment:
This part wasn't retrofitted to handle the compression operation and algorithm being added to `BlobInfo`s, so "replace with a compressed version" or "replace with an uncompressed version" is pretty much all it knows.

Reply:
I’m not quite sure what precision of options we want to expose here… For comparison, `types.SystemContext.CompressionFormat` only affects compression that happens during upload, `TryReusingBlob` is not affected.

Reply:
All the _existing_ users only really use this as a boolean — only use exact matches or “allow replacing with compressed cached versions”; no caller passes `types.Decompress`. i.e. se could just have an `allowUsingCompressedReplacements` boolean rather than passing `types.LayerCompression`.

---

`types.Decompress` could, within the current uses of the cache, only really make sense in pulls, if the cache is longer-lived than c/storage itself; but in that case, substituting the decompressed blob via `LayerInfosForCopy` would trigger a manifest update, and break digested references to the c/storage-stored outcome of the pull. So that case, at least, seems irrelevant or even something we outright should make impossible (within how digested references to c/storage work right now — I could see an argument that _that’s_ what should change instead).

That leaves completely third-party uses of the cache — pulling from a registry to `docker-archive:` or something like that. We neither use that nor exercise that in tests, so it seems to me that it would be better not to expose the “decompress” option at all.

(Eventually, for explicit Zstd support and fallbacks, that seems like something that should be driven by the copy pipeline, via extended `TryReusingBlob` etc. — those decisions are conceptually per-image, not per-cache, and should be consistent with what the copy pipeline does in `copyBlobFromStream`. I.e. that suggests extensions to the `ImageSource`/`ImageDestination`, rather than more options passed to this `ImageReference` wrapper.)